lib/vrl/stdlib/src/chunks.rs

```diff
@@ -0,0 +1,221 @@
+                required: true,
+            },
+            Parameter {
+                keyword: "utf8",
+                kind: kind::BOOLEAN,
+                required: true,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "chunks by byte when utf8=true",
+                source: r#"chunks("abcdefgh", 4, true)"#,
+                result: Ok(r#"["abcd", "efgh"]"#),
+            },
+            Example {
+                title: "chunk sizes respect unicode code point boundaries when utf8=true",
+                source: r#"chunks("ab你好", 4, true)"#,
+                result: Ok(r#"["ab", "你", "好"]"#),
+            },
+            Example {
+                title: "chunk size cannot be less than 4 bytes when utf8=true",
+                source: r#"chunks("hello world", 3, true)"#,
+                result: Err(
+                    r#"function call error for "chunks" at (0:30): "chunk_size" must be at least 4 bytes"#,
+                ),
+            },
+            Example {
+                title: "chunk size can be less than 4 bytes when utf8=false",
+                source: r#"chunks("hello world", 3, false)"#,
+                result: Ok(r#"["hel","lo ","wor","ld"]"#),
+            },
+            Example {
+                title: "chunk size must be at least 1 byte",
+                source: r#"chunks("hello world", 0, false)"#,
+                result: Err(
+                    r#"function call error for "chunks" at (0:31): "chunk_size" must be at least 1 byte"#,
+                ),
+            },
+            Example {
+                title: "chunk sizes do not respect unicode code point boundaries when utf8=false",
+                source: r#"chunks("ab你好", 4, false)"#,
+                result: Ok(r#"["ab�","�好"]"#),
+            },
+        ]
+    }
+
+    fn compile(
+        &self,
+        _state: (&mut state::LocalEnv, &mut state::ExternalEnv),
+        _ctx: &mut FunctionCompileContext,
+        mut arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+        let chunk_size = arguments.required("chunk_size");
+        let utf8 = arguments.required("utf8");
+
+        Ok(Box::new(ChunksFn {
+            value,
+            chunk_size,
+            utf8,
+        }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ChunksFn {
+    value: Box<dyn Expression>,
+    chunk_size: Box<dyn Expression>,
+    utf8: Box<dyn Expression>,
+}
+
+impl Expression for ChunksFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let chunk_size = self.chunk_size.resolve(ctx)?;
+        let utf8 = self.utf8.resolve(ctx)?;
+
+        chunks(value, chunk_size, utf8)
+    }
+
+    fn type_def(&self, _state: (&state::LocalEnv, &state::ExternalEnv)) -> TypeDef {
+        TypeDef::array(Collection::from_unknown(Kind::bytes())).infallible()
```

Comment:
This function is fallible, for example if the chunk size is set too low.
```return Err(r#""chunk_size" must be at least 1 byte"#.into());```

Since it's likely that a literal number will be passed into the function, we can be a bit smarter though. You can call `Expression::as_value` on the `chunk_size` argument, and this will resolve to a value at compile time if it can be calculated (if it's a literal). If so, you can do the length check at compile time and if it's >=1, return `infallible` here. Otherwise, it needs to be `fallible` since it could potentially fail.

You can assume the argument types will be correct when calculating fallibility here, since that happens automatically by the compiler.

Reply:
What happens when you use `Expression:as_value` on `chunk_size` but the `chunk_size` parameter is not a literal?

Is marking it as fallible or infallible just for documentation purposes? i.e. what are the ramifications of incorrectly annotating its fallibility?

Reply:
>What happens when you use Expression:as_value on chunk_size but the chunk_size parameter is not a literal?

`as_value()` returns an `Option<Value>`. It will return `None` if the value is not known at compile time.

> Is marking it as fallible or infallible just for documentation purposes? i.e. what are the ramifications of incorrectly annotating its fallibility?

This affects the type definition of the function, which is used at compile time. Fallible functions must have some kind of error handling.

I'd suggest running `vector vrl` if you haven't already to get access to the VRL REPL. You can play around with the function there and see how changing the type definition affects things.

Reply:
> I'd suggest running `vector vrl` if you haven't already to get access to the VRL REPL. You can play around with the function there and see how changing the type definition affects things.

Oh that sounds cool, I'll try that, thanks! 👍

Reply:
Oh okay, after playing with the REPL everything makes a whole lot more sense 😅 thanks again!

I'm trying to make it infallible with code like this:

```diff
diff --git a/lib/vrl/stdlib/src/chunks.rs b/lib/vrl/stdlib/src/chunks.rs
index 88afb3826..5872413bf 100644
--- a/lib/vrl/stdlib/src/chunks.rs
+++ b/lib/vrl/stdlib/src/chunks.rs
@@ -5,10 +5,6 @@ fn chunks(value: Value, chunk_size: Value) -> Resolved {
     let chunk_size = chunk_size.try_integer()? as usize;
     let bytes = value.try_bytes()?;
 
-    if chunk_size < 1 {
-        return Err(r#""chunk_size" must be at least 1 byte"#.into());
-    }
-
     Ok(bytes.chunks(chunk_size).collect::<Vec<_>>().into())
 }
 
@@ -66,6 +62,17 @@ impl Function for Chunks {
         let value = arguments.required("value");
         let chunk_size = arguments.required("chunk_size");
 
+        if let Some(chunk_size) = chunk_size.as_value() {
+            if chunk_size < 1 {
+                return Err(vrl::function::Error::InvalidArgument {
+                    keyword: "chunk_size",
+                    value: chunk_size,
+                    error: "Invalid chunk_size",
+                }
+                .into());
+            };
+        };
+
         Ok(Box::new(ChunksFn { value, chunk_size }))
     }
 }
```

But I'm getting this compiler error:

```
   Compiling vrl-stdlib v0.1.0 (/Users/brian/Code/github/vectordotdev/vector/lib/vrl/stdlib)
error[E0308]: mismatched types
  --> lib/vrl/stdlib/src/chunks.rs:66:29
   |
66 |             if chunk_size < 1 {
   |                             ^ expected enum `value::Value`, found integer
   |
help: try wrapping the expression in `value::Value::Integer`
   |
66 |             if chunk_size < value::Value::Integer(1) {
   |                             ++++++++++++++++++++++ +

For more information about this error, try `rustc --explain E0308`.
error: could not compile `vrl-stdlib` due to previous error
```
I'd like to be able to mark this function as infallible, but I'm not sure where to go from here (and I've got an errand to run). If you could provide some guidance on how to compare the Value to the integer, that would help! Otherwise I may just mark the function as fallible.

Reply:
> if let Some(chunk_size) = chunk_size.as_value() {

This gives you a `Value`. A `Value` is an enum of the different runtime types. You can check if the actual argument is an integer / get that integer by calling `as_integer` which returns `Option<i64>`, and then compare that value to 1.

Reply:
You will still want to keep the runtime check for the `chunk_size` that you deleted in the diff. This is useful if the argument passed in is a variable, or something unknown at compile time (which could fail)

Reply:
@fuchsnj I spent a bit of time trying to make the function infallible - I haven't gotten it quite yet, but my current direction is to try to enforce that the expression is a literal and not an expression similar to [redact.rs#L101-L139](https://github.com/vectordotdev/vector/blob/master/lib/vrl/stdlib/src/redact.rs#L101-L139). Does that sound like the right way to go?

I'm also checking in `fn compile()` to see if the given value is `>=1`:

```rust
if let Some(cs_value) = chunk_size.as_value() {
    if let Some(cs_int) = cs_value.as_integer() {
        if cs_int < 1 {
            return Err(vrl::function::Error::InvalidArgument {
                keyword: "chunk_size",
                value: cs_value,
                error: r#""chunk_size" must be at least 1 byte"#,
            }
            .into());
        }
    }
};
```

Reply:
I'm also a bit torn on whether this should be fallible or infallible. More thoughts on that when I can, but input would be appreciated, as well.